### PR TITLE
feat(MangroveOrder): Allow re-use of dead, owned offers to save gas and provision

### PR DESCRIPTION
**Pull request by lnist**
TODO: add tests of reuse and verify guards.

### CODE REVIEW

File: src/strategies/MangroveOrder.sol [hunk 1]
```diff
@@ -264,18 +264,25 @@ contract MangroveOrder is Forwarder, IOrderLogic {
...
+    });
+    if (tko.offerId == 0) {
+      (res.offerId,) = _newOffer(args, msg.sender);
+    } else {
+      uint offerId = tko.offerId;
+      require(ownerData[olKey.hash()][offerId].owner == msg.sender, "AccessControlled/Invalid");
+      require(!MGV.offers(olKey, offerId).isLive(), "mgvOrder/offerAlreadyActive");
+      if (_updateOffer(args, offerId) == REPOST_SUCCESS) {
+        res.offerId = offerId;
+      }
```
Comment (jkrivine) [inline src/strategies/MangroveOrder.sol:284]: to be consistent with the rest of the code we should add:
```solidity
else { 
  res.offerId = 0 ;
}

File: src/strategies/interfaces/IOrderLogic.sol [hunk 2]
```diff
@@ -22,6 +22,7 @@ interface IOrderLogic {
     bool fillWants;
     bool restingOrder;
     uint expiryDate;
+    uint offerId;
```
Comment (jkrivine) [inline src/strategies/interfaces/IOrderLogic.sol:25]: missing natspec for `offerId`

**Review by jkrivine — CHANGES_REQUESTED** [review]
A few comments.
Also on the sdk, how hard is it to reuse old offerIds?

File: src/strategies/interfaces/IOrderLogic.sol (hunk 2 above)
Comment (lnist) [reply src/strategies/interfaces/IOrderLogic.sol:26]: Fixed

File: src/strategies/MangroveOrder.sol (hunk 1 above)
Comment (lnist) [reply src/strategies/MangroveOrder.sol:286]: Added - though it is redundant initialization.

**Comment by lnist** [comment]
> A few comments. Also on the sdk, how hard is it to reuse old offerIds?

Via RPC we can filter on this event `event NewOwnedOffer(bytes32 indexed olKeyHash, uint indexed offerId, address indexed owner);` on olKeyHash+owner to get offers for that owner, and then check each if they are live.
Indexer can hand out all offers for owner.

**Pull request merged**